lib/react.dart

```diff
@@ -177,16 +177,21 @@ abstract class Component {
   /// Also allows [newState] to be used as a transactional `setState` callback.
   ///
   /// See: <https://facebook.github.io/react/docs/react-component.html#setstate>
-  void setState(dynamic newState, [callback()]) {
+  void setState(dynamic newState, [SetStateCallback callback]) {
     if (newState is Map) {
       _nextState.addAll(newState);
-    } else if (newState is _TransactionalSetStateCallback) {
+    } else if (newState is TransactionalSetStateCallback) {
       _transactionalSetStateCallbacks.add(newState);
     } else if (newState != null) {
-      throw new ArgumentError('setState expects its first parameter to either be a Map or a Function that accepts two parameters.');
+      throw new ArgumentError('setState expects its first parameter to either be a Map or a `TransactionalSetStateCallback`.');
     }
 
-    if (callback != null) _setStateCallbacks.add(callback);
+    if (callback != null) {
+      // Store locally to prevent concurrent modification during iteration
+      var localSetStateCallbacks = new List<SetStateCallback>.from(_setStateCallbacks);
+      localSetStateCallbacks.add(callback);
+      _setStateCallbacks = localSetStateCallbacks;
```

Comment:
Looks like this is necessary for `replaceState` as well. You could also prevent the concurrent modification when the callbacks are iterated, which would require no changes to either of these:
```dart
  void _callSetStateCallbacks(Component component) {
    var callbacks = component.setStateCallbacks.toList();
    component.setStateCallbacks.clear();
    callbacks.forEach((callback()) { callback(); });
  }
```